### ISSUE TITLE
2024.05.2 breaks my camera

### ISSUE THREAD

**Issue by nomonkeynodeal** [sca075/valetudo_vacuum_camera#140]
### Checklist

- [X] I have updated the integration to the latest version available
- [X] I have checked if the problem is already reported

### The problem

2024.05.2 breaks my camera. Everything works again after rolling back to 2024.05.1.

The cards just say "valetudo invalid calibration, please check your configuration". Two browsers, cleared cache.

I set this up a while ago and barely remember how anything works, much less how to troubleshoot. If anything sticks out, let me know. Otherwise I guess I will just keep this version until the heatdeath of the universe.

### What version of an integration has described problem?

2024.05.2

### What was the last working version of an integration?

2024.05.1

### What vacuum model do you have problems with?

Dreame L10 Pro

### Please firmware installed on your Vacuum.

2023.12.0

### What type of platform you use?

VEMLinux (Virtual Machine) such as Proxmox, VMWare, etc.

### What version of Home Assistant do you use?

latest as of today

### What type of installation are you running?

Home Assistant OS

### Camera's settings.

```yaml
- type: custom:xiaomi-vacuum-map-card
        map_source:
          camera: camera.dreame_l10_2_camera
        calibration_source:
          camera: true
        entity: vacuum.valetudo_l10_2
        vacuum_platform: Hypfer/Valetudo
        tiles:
          - tile_id: battery_level
            entity: vacuum.valetudo_l10_2
            label: Battery
            attribute: battery_level
            icon_source: vacuum.valetudo_l10_2.attributes.battery_icon
            unit: '%'
          - tile_id: fan_speed
            entity: vacuum.valetudo_l10_2
            label: Fan speed
            attribute: fan_speed
            icon: mdi:fan
            translations:
              silent: Silent
              standard: Standard
              medium: Medium
              turbo: Turbo
              auto: Auto
              gentle: Gentle
          - tile_id: filter_left
            label: Filter left
            icon: mdi:air-filter
            unit: h
            multiplier: 0.016666666666666666
            entity: sensor.valetudo_l10_2_main_filter
            precision: 0
            translations: {}
          - tile_id: main_brush_left
            label: Main brush left
            icon: mdi:brush
            unit: h
            multiplier: 0.016666666666666666
            entity: sensor.valetudo_l10_2_main_brush
            precision: 0
            translations: {}
          - tile_id: side_brush_left
            label: Side brush left
            icon: mdi:brush
            unit: h
            multiplier: 0.016666666666666666
            entity: sensor.valetudo_l10_2_right_brush
            precision: 0
            translations: {}
          - tile_id: cleaned_area
            label: Cleaned area
            icon: mdi:texture-box
            unit: m²
            multiplier: 0.0001
            entity: sensor.valetudo_l10_2_current_statistics_area
            precision: 0
            translations: {}
          - tile_id: cleaning_time
            label: Cleaning time
            icon: mdi:timer-sand
            unit: min
            multiplier: 0.016666666666666666
            entity: sensor.valetudo_l10_2_current_statistics_time
            precision: 0
            translations: {}
        icons: []
        map_locked: true
        two_finger_pan: true
        map_modes:
          - template: vacuum_clean_segment
            service_call_schema:
              service: mqtt.publish
              evaluate_data_as_template: false
              service_data:
                topic: valetudo/L10-2/MapSegmentationCapability/clean/set
                payload: >-
                  {"segment_ids": [[selection]], "iterations": [[repeats]],
                  "customOrder": false}   
            predefined_selections:
              - id: '1'
                icon:
                  name: mdi:shower-head
                  x: 3102
                  'y': 2962
                label:
                  text: Office bath
                  x: 3102
                  'y': 2962
                  offset_y: 35
                outline:
                  - - 2980
                    - 2875
                  - - 3225
                    - 2875
                  - - 3225
                    - 3050
                  - - 2980
                    - 3050
              - id: '2'
                icon:
                  name: mdi:shower-head
                  x: 2735
                  'y': 2985
                label:
                  text: Master bath
                  x: 2735
                  'y': 2985
                  offset_y: 35
                outline:
                  - - 2485
                    - 2885
                  - - 2985
                    - 2885
                  - - 2985
                    - 3085
                  - - 2485
                    - 3085
              - id: '4'
                icon:
                  name: mdi:chair-rolling
                  x: 3142
                  'y': 3305
                label:
                  text: Office
                  x: 3142
                  'y': 3305
                  offset_y: 35
                outline:
                  - - 2985
                    - 3055
                  - - 3300
                    - 3055
                  - - 3300
                    - 3555
                  - - 2985
                    - 3555
              - id: '5'
                icon:
                  name: mdi:numeric-3-box
                  x: 2255
                  'y': 3465
                label:
                  text: Hall
                  x: 2255
                  'y': 3465
                  offset_y: 35
                outline:
                  - - 1830
                    - 3335
                  - - 2680
                    - 3335
                  - - 2680
                    - 3595
                  - - 1830
                    - 3595
              - id: '3'
                icon:
                  name: mdi:bed-queen
                  x: 2200
                  'y': 3117
                label:
                  text: Master bed
                  x: 2200
                  'y': 3117
                  offset_y: 35
                outline:
                  - - 2020
                    - 2885
                  - - 2380
                    - 2885
                  - - 2380
                    - 3350
                  - - 2020
                    - 3350
          - name: Zone cleanup
            icon: mdi:select-drag
            run_immediately: false
            coordinates_rounding: true
            coordinates_to_meters_divider: 100
            selection_type: MANUAL_RECTANGLE
            max_selections: 5
            repeats_type: EXTERNAL
            max_repeats: 3
            service_call_schema:
              service: mqtt.publish
              evaluate_data_as_template: true
              service_data:
                topic: valetudo/L10-2/ZoneCleaningCapability/start/set
                payload: >-
                  {"zones": [{%for s in ('[[selection]]')|from_json %}{
                  "points": {"pA": { "x": {{s[0]}}, "y": {{s[1]}} }, "pB": {
                  "x": {{s[2]}}, "y": {{s[1]}} }, "pC": { "x": {{s[2]}}, "y":
                  {{s[3]}} }, "pD": { "x": {{s[0]}}, "y": {{s[3]}} } },
                  "iterations": [[repeats]]}{%if not
                  loop.last%},{%endif%}{%endfor%}]}
            predefined_selections: []
            variables: {}
```

**Comment by sca075** [comment]
@nomonkeynodeal is there a warning that the vacuum is not connected?
I would need the logs of the camera not the configuration of the card.

**Comment by brusarp** [comment]
`Update for camera.fuffy_camera fails
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 950, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1293, in async_device_update
    await self.async_update()
  File "/config/custom_components/valetudo_vacuum_camera/camera.py", line 425, in async_update
    pil_img = await self.hass.async_create_task(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/valetudo_vacuum_camera/camera_processing.py", line 191, in run_async_process_valetudo_data
    images = await gather(*tasks)
             ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/valetudo_vacuum_camera/camera_processing.py", line 169, in process_valetudo_data
    result = loop.run_until_complete(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 685, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/config/custom_components/valetudo_vacuum_camera/camera_processing.py", line 48, in async_process_valetudo_data
    pil_img = await self._map_handler.async_get_image_from_json(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/valetudo_vacuum_camera/valetudo/hypfer/image_handler.py", line 268, in async_get_image_from_json
    room_id, img_np_array = await self.imd.async_draw_base_layer(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/valetudo_vacuum_camera/valetudo/hypfer/image_draw.py", line 68, in async_draw_base_layer
    and self.img_h.active_zones[room_id + 1] == 1
        ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
IndexError: list index out of range
`

Mine too, is enough this log?

**Comment by sca075** [comment]
@brusarp Will have a look on this as soon. It was somehow expected as I did modify the Image_handler.py thanks a lot for the report ;)

**Comment by sca075** [comment]
@nomonkeynodeal I noticed that when the vacuum get disconnected this is going to happen, is your problem as this?
![Screenshot 2024-05-11 at 12 44 26](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/42a4a6db-5909-4d17-a74e-0dce1ae294e0)

If yes I will try to fix it within the next version. Thanks also for you report.

**Comment by sca075** [comment]
@nomonkeynodeal  I did check and probably found the issue in the connector.py that was causing the camera to stop.
The correction for this issue, if it is the same you are facing (as I didn't yet got the logs of the camera) this was a quick easy and easy fix. Dreame are having some issue with MQTT (firmware wise) and also.. 1138 isn't actually a firmware I'm familiar with. I mean what Valetudo version are you using?

**Comment by brusarp** [comment]
If I can be helpful...
Mine is Valetudo 2024.02.0

![Screenshot_20240511_134101_Chrome.jpg](https://github.com/sca075/valetudo_vacuum_camera/assets/14950532/1908db8e-8637-40ce-8557-518c75ff166e)

**Comment by sca075** [comment]
> If I can be helpful... Mine is Valetudo 2024.02.0 Dreame D9
> 
> ![Screenshot_20240511_134101_Chrome.jpg](https://private-user-images.githubusercontent.com/14950532/329778571-1908db8e-8637-40ce-8557-518c75ff166e.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU0MjgxNjEsIm5iZiI6MTcxNTQyNzg2MSwicGF0aCI6Ii8xNDk1MDUzMi8zMjk3Nzg1NzEtMTkwOGRiOGUtODYzNy00MGNlLTg1NTctNTE4Yzc1ZmYxNjZlLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTExVDExNDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhMGI5NjQyYWFjZmZjZWU2Y2FjYmEzYWU5NWIzMTJjNjVhYjc0YjFjY2RlYTkwNDUwY2I5MDFjODNjOGVlOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.h6mCMu92HZqUEtJssGr5kqUc-rB4_JG_YnpqRkpJXQ4)

Yep this help thanks @brusarp .. @nomonkeynodeal can you please confirm it is the same for you?

**Comment by benbender** [comment]
@sca075 can confirm. Seeing the same error after updating.

**Comment by nomonkeynodeal** [comment]
> @nomonkeynodeal I did check and probably found the issue in the connector.py that was causing the camera to stop. The correction for this issue, if it is the same you are facing (as I didn't yet got the logs of the camera) this was a quick easy and easy fix. Dreame are having some issue with MQTT (firmware wise) and also.. 1138 isn't actually a firmware I'm familiar with. I mean what Valetudo version are you using?

Sorry, that is the manufacturer firmware version. I was using Valetudo version 2023.12.0, but I upgraded to 2024.2.0 just now.

I looked for anything in the HA logs before rolling back to 2024.5.1, but I couldn't find any errors. Maybe I needed to look harder. I'd happy to test out whatever you changed.

**Comment by sca075** [comment]
All, thanks a lot for sharing your experience and feedback.

I'm working on it and probably will release a fix as soon.

I will test the camera with a Drame JSON and come back to you as soon I'm done with it 😉

**Comment by sca075** [comment]
Okay I was able to reproduce the issue #142 by loading the test json of a D9Pro I had, the condition failing was changed to:
```python
if layer_type == "segment":
                    # Check if the room is active and set a modified color
                    if (
                        self.img_h.active_zones
                        and room_id < (len(self.img_h.active_zones) - 1)
                    ):
                        if self.img_h.active_zones[room_id + 1] == 1:
                            room_color = (
                                ((2 * room_color[0]) + color_zone_clean[0]) // 3,
                                ((2 * room_color[1]) + color_zone_clean[1]) // 3,
                                ((2 * room_color[2]) + color_zone_clean[2]) // 3,
                                ((2 * room_color[3]) + color_zone_clean[3]) // 3,
                            )
```
The original code lines are: https://github.com/sca075/valetudo_vacuum_camera/blob/ae25662dce40e0f85d12540e506652c29f535187/custom_components/valetudo_vacuum_camera/valetudo/hypfer/image_draw.py#L62-L74
 
I need some additional test to confirm it works as we expect and as soon I'm done with it I will release the fixes for both connector.py and image_draw.py, once again thanks for your kind understanding and cooperation guys.
If you would consider to star the project this would help :)

**Comment by sca075** [comment]
All just released the 2024.05.3 that will fix the issues you are facing. At your complete disposal for any clarification's or other requests. Thanks once again for your contributions :)

_If you would consider to **star** the project it would be really appreciated._

**Comment by nomonkeynodeal** [comment]
> Okay I was able to reproduce the issue #142 by loading the test json of a D9Pro I had, the condition failing was changed to:
> 
> ```python
> if layer_type == "segment":
>                     # Check if the room is active and set a modified color
>                     if (
>                         self.img_h.active_zones
>                         and room_id < (len(self.img_h.active_zones) - 1)
>                     ):
>                         if self.img_h.active_zones[room_id + 1] == 1:
>                             room_color = (
>                                 ((2 * room_color[0]) + color_zone_clean[0]) // 3,
>                                 ((2 * room_color[1]) + color_zone_clean[1]) // 3,
>                                 ((2 * room_color[2]) + color_zone_clean[2]) // 3,
>                                 ((2 * room_color[3]) + color_zone_clean[3]) // 3,
>                             )
> ```
> 
> The original code lines are:
> 
> https://github.com/sca075/valetudo_vacuum_camera/blob/ae25662dce40e0f85d12540e506652c29f535187/custom_components/valetudo_vacuum_camera/valetudo/hypfer/image_draw.py#L62-L74
> 
> I need some additional test to confirm it works as we expect and as soon I'm done with it I will release the fixes for both connector.py and image_draw.py, once again thanks for your kind understanding and cooperation guys. If you would consider to star the project this would help :)

Thanks, I can confirm it's working!
